In [68]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### Reading and Accessing the data

In [98]:
df = pd.read_csv('Data_for_UCI_named.csv')
df.head()

tau1      tau2      tau3      tau4        p1        p2        p3  \
0  2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1  9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2  8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3  0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4  3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   

         p4        g1        g2        g3        g4      stab     stabf  
0 -1.723086  0.650456  0.859578  0.887445  0.958034  0.055347  unstable  
1 -1.255012  0.413441  0.862414  0.562139  0.781760 -0.005957    stable  
2 -0.920492  0.163041  0.766689  0.839444  0.109853  0.003471  unstable  
3 -0.997374  0.446209  0.976744  0.929381  0.362718  0.028871  unstable  
4 -0.554305  0.797110  0.455450  0.656947  0.820923  0.049860  unstable

In [99]:
df.shape

(10000, 14)

In [100]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   tau1    10000 non-null  float64
 1   tau2    10000 non-null  float64
 2   tau3    10000 non-null  float64
 3   tau4    10000 non-null  float64
 4   p1      10000 non-null  float64
 5   p2      10000 non-null  float64
 6   p3      10000 non-null  float64
 7   p4      10000 non-null  float64
 8   g1      10000 non-null  float64
 9   g2      10000 non-null  float64
 10  g3      10000 non-null  float64
 11  g4      10000 non-null  float64
 12  stab    10000 non-null  float64
 13  stabf   10000 non-null  object 
dtypes: float64(13), object(1)
memory usage: 1.1+ MB


In [101]:
df.duplicated().sum()

0

In [102]:
df.describe()

tau1          tau2          tau3          tau4            p1  \
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.000000   
mean       5.250000      5.250001      5.250004      5.249997      3.750000   
std        2.742548      2.742549      2.742549      2.742556      0.752160   
min        0.500793      0.500141      0.500788      0.500473      1.582590   
25%        2.874892      2.875140      2.875522      2.874950      3.218300   
50%        5.250004      5.249981      5.249979      5.249734      3.751025   
75%        7.624690      7.624893      7.624948      7.624838      4.282420   
max        9.999469      9.999837      9.999450      9.999443      5.864418   

                 p2            p3            p4            g1            g2  \
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.000000   
mean      -1.250000     -1.250000     -1.250000      0.525000      0.525000   
std        0.433035      0.433035      0.433035      0.274256      0.274255   
min       -1.999891     -1.999945     -1.999926      0.050009      0.050053   
25%       -1.624901     -1.625025     -1.624960      0.287521      0.287552   
50%       -1.249966     -1.249974     -1.250007      0.525009      0.525003   
75%       -0.874977     -0.875043     -0.875065      0.762435      0.762490   
max       -0.500108     -0.500072     -0.500025      0.999937      0.999944   

                 g3            g4          stab  
count  10000.000000  10000.000000  10000.000000  
mean       0.525000      0.525000      0.015731  
std        0.274255      0.274255      0.036919  
min        0.050054      0.050028     -0.080760  
25%        0.287514      0.287494     -0.015557  
50%        0.525015      0.525002      0.017142  
75%        0.762440      0.762433      0.044878  
max        0.999982      0.999930      0.109403

In [103]:
df.corr()

tau1      tau2      tau3      tau4        p1        p2        p3  \
tau1  1.000000  0.015586 -0.005970 -0.017265  0.027183 -0.015485 -0.015924   
tau2  0.015586  1.000000  0.014273 -0.001965 -0.004769  0.006573  0.007673   
tau3 -0.005970  0.014273  1.000000  0.004354  0.016953 -0.003134 -0.008780   
tau4 -0.017265 -0.001965  0.004354  1.000000 -0.003173  0.010553  0.006169   
p1    0.027183 -0.004769  0.016953 -0.003173  1.000000 -0.573157 -0.584554   
p2   -0.015485  0.006573 -0.003134  0.010553 -0.573157  1.000000  0.002388   
p3   -0.015924  0.007673 -0.008780  0.006169 -0.584554  0.002388  1.000000   
p4   -0.015807 -0.005963 -0.017531 -0.011211 -0.579239 -0.006844  0.012953   
g1    0.010521 -0.001742 -0.011605 -0.004149  0.000721  0.015603 -0.003219   
g2    0.015350  0.015383  0.007671  0.008431  0.015405 -0.018032 -0.011575   
g3   -0.001279  0.016508  0.014702  0.003260  0.001069  0.007555 -0.005897   
g4    0.005494 -0.011764 -0.011497 -0.000491 -0.015451  0.019817 -0.010485   
stab  0.275761  0.290975  0.280700  0.278576  0.010278  0.006255 -0.003321   

            p4        g1        g2        g3        g4      stab  
tau1 -0.015807  0.010521  0.015350 -0.001279  0.005494  0.275761  
tau2 -0.005963 -0.001742  0.015383  0.016508 -0.011764  0.290975  
tau3 -0.017531 -0.011605  0.007671  0.014702 -0.011497  0.280700  
tau4 -0.011211 -0.004149  0.008431  0.003260 -0.000491  0.278576  
p1   -0.579239  0.000721  0.015405  0.001069 -0.015451  0.010278  
p2   -0.006844  0.015603 -0.018032  0.007555  0.019817  0.006255  
p3    0.012953 -0.003219 -0.011575 -0.005897 -0.010485 -0.003321  
p4    1.000000 -0.013636  0.002850 -0.003515  0.017505 -0.020786  
g1   -0.013636  1.000000  0.007559 -0.005836  0.012431  0.282774  
g2    0.002850  0.007559  1.000000 -0.012809 -0.014909  0.293601  
g3   -0.003515 -0.005836 -0.012809  1.000000  0.006900  0.308235  
g4    0.017505  0.012431 -0.014909  0.006900  1.000000  0.279214  
stab -0.020786  0.282774  0.293601  0.308235  0.279214  1.000000

In [104]:
# Drop the stab column since it's highly correlated with the target 
df.drop('stab', axis = 1, inplace=True)

In [105]:
#test
assert 'stab' not in df.columns

### Feature and Target variable creation and spliting

In [106]:
# Percentage of the various classes in the target variable
df.stabf.value_counts()/df.stabf.count() * 100

unstable    63.8
stable      36.2
Name: stabf, dtype: float64


We have an imbalance classes but sufficeint data for each class so we simply use stratified split to maintain class ratio for 
between train and test split

In [107]:
# Create feature and target arrays
X = df.drop('stabf', axis = 1)
y = df.stabf
print(X.shape)
print(y.shape)

(10000, 12)
(10000,)


In [108]:
### we carryout a stratified split of  the data with a test_size of 0.2 and random state of 1.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.2, random_state= 1, stratify= y)

In [109]:
# Confirm classes percentage of original dataset were retained for train and test y
y_train.value_counts()/y_train.count() * 100

unstable    63.8
stable      36.2
Name: stabf, dtype: float64

In [110]:
y_test.value_counts()/y_test.count() * 100

unstable    63.8
stable      36.2
Name: stabf, dtype: float64

### Features Scaling

In [145]:

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
x_train = scaler.fit_transform(X_train)
x_test = scaler.fit_transform(X_test)

### Random Forest

In [146]:
rf = RandomForestClassifier()
# Train model useing our train set and make prediction using test set
rf.fit(x_train, y_train)
y_pred = rf.predict(x_test)
# check for model accuracy
r_accuracy = accuracy_score(y_test, y_pred)
print(f'accuracy = {r_accuracy}')

accuracy = 0.9255


In [56]:
# import, instantiate and train model
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score


def train_predict_rf(n_est, max_depth):
    """
    The function train a random forest model for various hyperparameter combination, predict values for 
    the target variable and print out a classification report.
    INPUTS:
    n_est: integer value for the number of tree of a random forest model ie n_estimator
    max_depth: integer value for the tree depth of a random forest model

    """
    #initialize model
    rf = RandomForestClassifier(n_estimators= n_est, max_features= 'auto', random_state= 1, max_depth= max_depth)
    # Train model useing our train set and make prediction using test set
    rf.fit(x_train, y_train)
    y_pred = rf.predict(x_test)
    # check for model accuracy
    r_accuracy = accuracy_score(y_test, y_pred)
    print(f'accuracy = {r_accuracy}')
    # print out classification report
    rf_report = classification_report(y_test, y_pred, zero_division = 0)
    print(rf_report)

In [60]:
# run train_predict_rf for different values of n_estimators to tune for the number of trees
n_trees = [50, 200, 500, 1000]
for num in n_trees:
    print('For a n_estimator of {} while max_depth is 80'.format(num))
    train_predict_rf(n_est= num, max_depth= 80)

For a n_estimator of 50 while max_depth is 80
accuracy = 0.9185
              precision    recall  f1-score   support

      stable       0.90      0.87      0.89       724
    unstable       0.93      0.95      0.94      1276

    accuracy                           0.92      2000
   macro avg       0.91      0.91      0.91      2000
weighted avg       0.92      0.92      0.92      2000

For a n_estimator of 200 while max_depth is 80
accuracy = 0.9235
              precision    recall  f1-score   support

      stable       0.92      0.86      0.89       724
    unstable       0.93      0.96      0.94      1276

    accuracy                           0.92      2000
   macro avg       0.92      0.91      0.92      2000
weighted avg       0.92      0.92      0.92      2000

For a n_estimator of 500 while max_depth is 80
accuracy = 0.9245
              precision    recall  f1-score   support

      stable       0.92      0.86      0.89       724
    unstable       0.92      0.96      0.94

In [61]:
n_trees = [300, 600, 900]
for num in n_trees:
    print('For a n_estimator of {} while max_depth is 80'.format(num))
    train_predict_rf(n_est= num, max_depth= 80)

For a n_estimator of 300 while max_depth is 80
accuracy = 0.9255
              precision    recall  f1-score   support

      stable       0.92      0.86      0.89       724
    unstable       0.93      0.96      0.94      1276

    accuracy                           0.93      2000
   macro avg       0.93      0.91      0.92      2000
weighted avg       0.93      0.93      0.92      2000

For a n_estimator of 600 while max_depth is 80
accuracy = 0.922
              precision    recall  f1-score   support

      stable       0.92      0.86      0.89       724
    unstable       0.92      0.96      0.94      1276

    accuracy                           0.92      2000
   macro avg       0.92      0.91      0.91      2000
weighted avg       0.92      0.92      0.92      2000

For a n_estimator of 900 while max_depth is 80
accuracy = 0.9225
              precision    recall  f1-score   support

      stable       0.92      0.86      0.89       724
    unstable       0.92      0.96      0.94

We have the best metric when n_estimator is set to 300. An accuracy of 0.9255, precision of 0.92 for both the stable and unstable class, recall of 0.86 and 0.96 for the stable and unstable class respectively and f1-score of 0.89 and 0.94 for the stable and unstable classes respectively.

In [65]:
# Using a n_estimator of 300, we run train_predict_rf for different values of max_depth
n_depth = [100, 200, 500, 1000]
for depth in n_depth:
    print('For a n_estimator of 300 while max_depth is {}'.format(depth))
    train_predict_rf(n_est= 300, max_depth= depth)

For a n_estimator of 300 while max_depth is 100
accuracy = 0.9255
              precision    recall  f1-score   support

      stable       0.92      0.86      0.89       724
    unstable       0.93      0.96      0.94      1276

    accuracy                           0.93      2000
   macro avg       0.93      0.91      0.92      2000
weighted avg       0.93      0.93      0.92      2000

For a n_estimator of 300 while max_depth is 200
accuracy = 0.9255
              precision    recall  f1-score   support

      stable       0.92      0.86      0.89       724
    unstable       0.93      0.96      0.94      1276

    accuracy                           0.93      2000
   macro avg       0.93      0.91      0.92      2000
weighted avg       0.93      0.93      0.92      2000

For a n_estimator of 300 while max_depth is 500
accuracy = 0.9255
              precision    recall  f1-score   support

      stable       0.92      0.86      0.89       724
    unstable       0.93      0.96      

Metrics remains same irrespective of tree depth.

### Extra Tree Classifier

In [147]:
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.metrics import accuracy_score
ext = ExtraTreesClassifier(random_state= 1)
ext.fit(x_train, y_train)
y_pred = ext.predict(x_test)
acc = accuracy_score(y_test, y_pred)
print(acc)

0.928


In [148]:
report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

      stable       0.94      0.85      0.90       724
    unstable       0.92      0.97      0.94      1276

    accuracy                           0.93      2000
   macro avg       0.93      0.91      0.92      2000
weighted avg       0.93      0.93      0.93      2000



In [153]:
ext.feature_importances_

array([0.11549494, 0.11994856, 0.11209228, 0.11901472, 0.03979815,
       0.04130391, 0.04139812, 0.04112188, 0.08948417, 0.09420962,
       0.09738701, 0.08874665])

In [151]:
X_train.columns

Index(['tau1', 'tau2', 'tau3', 'tau4', 'p1', 'p2', 'p3', 'p4', 'g1', 'g2',
       'g3', 'g4'],
      dtype='object')

In [81]:
def train_predict_ext(n_est, criterion, max_depth):
    """
    The function train an extra tree classifier model for various hyperparameter combination, predict values for 
    the target variable and print out model accuracy and  classification report.
    INPUTS:
    n_est: integer value for the number of tree of a random forest model ie n_estimator
    max_depth: integer value for the tree depth of a random forest model
    criterion: String - one of three possible values; gini, entropy and log_loss 
    """
    
    # Instantiate the model
    ext = ExtraTreesClassifier(n_estimators= n_est, criterion= criterion , max_depth= max_depth, random_state= 1)
    # Fit the model
    ext.fit(x_train, y_train)
    # Predict on test data
    y_pred = ext.predict(x_test)
    # check for model accuracy
    acc = accuracy_score(y_test, y_pred)
    print(f'accuracy = {acc}')
    # Display classification report
    report = classification_report(y_test, y_pred)
    print(report)
    

In [88]:
# We iterate for the various citerias keeping max_depth at 100 and n_estimator at 100
criteria = ['gini', 'entropy']
for criterion in criteria:
    print(f'For {criterion} with max_depth of 100 and n_est of 100')
    train_predict_ext(n_est= 100, criterion= criterion, max_depth= 100)
    

For gini with max_depth of 100 and n_est of 100
accuracy = 0.928
              precision    recall  f1-score   support

      stable       0.94      0.85      0.90       724
    unstable       0.92      0.97      0.94      1276

    accuracy                           0.93      2000
   macro avg       0.93      0.91      0.92      2000
weighted avg       0.93      0.93      0.93      2000

For entropy with max_depth of 100 and n_est of 100
accuracy = 0.922
              precision    recall  f1-score   support

      stable       0.94      0.84      0.89       724
    unstable       0.92      0.97      0.94      1276

    accuracy                           0.92      2000
   macro avg       0.93      0.90      0.91      2000
weighted avg       0.92      0.92      0.92      2000



In [92]:
# Using gini and a max_depth of 100, we evaluate for various n_estimator
n_trees = [200, 500, 1000, 1500, 2000]
for n in n_trees:
    print(f"For n_estimator of {n}, using gini with max_depth of 100")
    train_predict_ext(n_est= n, max_depth= 100, criterion = 'gini')

For n_estimator of 200, using gini with max_depth of 100
accuracy = 0.9275
              precision    recall  f1-score   support

      stable       0.95      0.85      0.89       724
    unstable       0.92      0.97      0.94      1276

    accuracy                           0.93      2000
   macro avg       0.93      0.91      0.92      2000
weighted avg       0.93      0.93      0.93      2000

For n_estimator of 500, using gini with max_depth of 100
accuracy = 0.931
              precision    recall  f1-score   support

      stable       0.96      0.85      0.90       724
    unstable       0.92      0.98      0.95      1276

    accuracy                           0.93      2000
   macro avg       0.94      0.91      0.92      2000
weighted avg       0.93      0.93      0.93      2000

For n_estimator of 1000, using gini with max_depth of 100
accuracy = 0.9325
              precision    recall  f1-score   support

      stable       0.96      0.85      0.90       724
    unstable

The best metric was obtained at n_estimator = 1500 and a max_depth of 100 using gini criterion. This gave an accuracy of 0.934, precison of 0.96 for the stable class and 0.92 for the unstable class with an f1 score of 0.9 and 0.95 for the stable and unstable classes respectively.

### Encoding of target variable

In [131]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
eny_train = le.fit_transform(y_train)
eny_test = le.transform(y_test)

In [132]:
def get_integer_mapping(le):
    '''
    Return a dict mapping labels to their integer values
    from an SKlearn LabelEncoder
    le = a fitted SKlearn LabelEncoder
    '''
    res = {}
    for cl in le.classes_:
        res.update({cl:le.transform([cl])[0]})

    return res

In [133]:
get_integer_mapping(le)

{'stable': 0, 'unstable': 1}

### XGBoost

In [134]:
! pip install xgboost

In [135]:
from xgboost import XGBClassifier
xgb = XGBClassifier()
xgb.fit(x_train, eny_train)
pred = xgb.predict(x_test)

In [136]:
xgb_report = classification_report(eny_test, pred)

In [137]:
print(xgb_report)

              precision    recall  f1-score   support

           0       0.94      0.92      0.93       724
           1       0.95      0.97      0.96      1276

    accuracy                           0.95      2000
   macro avg       0.95      0.94      0.94      2000
weighted avg       0.95      0.95      0.95      2000



In [138]:
accuracy = accuracy_score(xgy_test, pred)
print(accuracy)

0.948


### LightGBM

In [130]:
! pip install lightgbm

In [140]:
from lightgbm import LGBMClassifier
lgb = LGBMClassifier()
lgb.fit(x_train, eny_train)
pred = lgb.predict(x_test)

In [142]:
lgb_report = classification_report(eny_test, pred)
print(lgb_report)

              precision    recall  f1-score   support

           0       0.93      0.92      0.93       724
           1       0.95      0.96      0.96      1276

    accuracy                           0.95      2000
   macro avg       0.94      0.94      0.94      2000
weighted avg       0.95      0.95      0.95      2000



In [143]:
acc = accuracy_score(eny_test, pred)
print(acc)

0.9465
